### ISSUE TITLE
Improved category support for blog plugin

### ISSUE THREAD

**Issue by ZhiyuanChen** [squidfunk/mkdocs-material#4719]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I want to suggest an idea and checked that ...

- [X] ... to my best knowledge, my idea wouldn't break something for other users
- [X] ... the documentation does not mention anything about my idea
- [X] ... there are no open or closed issues that are related to my idea

### Description

1. `post_url_format` can only comprises `date`, `file`(file.name), `slug`. Allow categories/sections to be part of `post_url_format`
2. next/prev page searches globally, allow them only look for posts in the same categories/sections

### Use Cases

I have my blog built on MkDocs. 
Before the blog plugins, I have different sections for holding blogs with different topics, and it worked grate. 
Now, as the section is taken by author information, and next/prev page searches globally instead of inside section only, it's very hard to find other blogs in the same section. 
I can certainly use categories instead of sections, but it's still hard to find other posts easily.

**Comment by squidfunk** [comment]
Thanks for suggesting. Could you please illustrate your feature request with an example?

**Comment by ZhiyuanChen** [comment]
For example, [https://zyc.ai/统计/index/](https://zyc.ai/统计/index/) and are two blog page (sorry its in Chinese, but... the content does not matter
I cannot use `post_url_format: "{file}"` here because they are both index page, and it's a bit strange to have `post_url_format: "{date}/{file}"` here because it's not relevant to date. 
One solution is to use the categories page instead of using a specific index page, I believe it is actually preferred here. 
However, the section tab (table in the left) of categories page lists all categories, but I only want it to show all pages in this category. 
<img width="1072" alt="image" src="https://user-images.githubusercontent.com/28757366/206538001-47141acd-9964-466e-8f11-c927615019df.png">

And for this page [https://zyc.ai/clipclip/clip/#_10](https://zyc.ai/clipclip/clip/#_10), it is a paper reading page and everything other pages in the same categories/section are also paper readings. But its next and prev page is a sketch page.
<img width="1334" alt="image" src="https://user-images.githubusercontent.com/28757366/206538085-2354cc6e-6b54-424d-abc1-ea4ffef97b40.png">
I would really want the next/prev to be other paper/reading page. 

It would be even better to have a navigation section on the left or other pages to list other pages in the same section/categories. As it's really hard to locate other pages now.

**Comment by squidfunk** [comment]
Thanks for the additional information.

> However, the section tab (table in the left) of categories page lists all categories, but I only want it to show all pages in this category.

You can create a normal MkDocs project with a regular navigations structure then. The blog plugin makes no sense here, since the categories feature is intended to generate category index pages. As you mentioned, you don't want index pages, but a regular navigation with the actual pages. 

> [...] because it's not relevant to date.

This is another hint that the blog plugin might not be the best fit. Other than that, as I already mentioned, you can divide your documentation into multiple blogs per top-level category.

Closing as out-of-scope.

**Comment by ZhiyuanChen** [comment]
Actually I think an index page is a grate idea, but it needs to function as an index page and helps readers to index an article. 

Therefore, I think category index pages should be more category-centric, it makes more sense to list what articles are in this category rather than what are the rest of the categories. 

List all articles work for a tiny blog, but if there are 100 or even more articles inside one category, it is nearly impossible for readers to find a specific article. 

The functions specified in #4363 actually satisfy all my needs in terms of content.

**Comment by squidfunk** [comment]
> Therefore, I think category index pages should be more category-centric, it makes more sense to list what articles are in this category rather than what are the rest of the categories.

That is exactly what category index pages do – they list all articles that are assigned to the category. For example, the "Search" category [on our blog](https://squidfunk.github.io/mkdocs-material/blog/category/search/) lists all articles within that category:

![screenshot-squidfunk-github-io-mkdocs-material-blog-category-search-1670532555917](https://user-images.githubusercontent.com/932156/206564581-f2b92d9e-790a-4a5c-835c-fbd36ff3585c.png)

**Comment by ZhiyuanChen** [comment]
> > Therefore, I think category index pages should be more category-centric, it makes more sense to list what articles are in this category rather than what are the rest of the categories.
> 
> That is exactly what category index pages do – they list all articles that are assigned to the category. For example, the "Search" category [on our blog](https://squidfunk.github.io/mkdocs-material/blog/category/search/) lists all articles within that category:
> 
> ![screenshot-squidfunk-github-io-mkdocs-material-blog-category-search-1670532555917](https://user-images.githubusercontent.com/932156/206564581-f2b92d9e-790a-4a5c-835c-fbd36ff3585c.png)

Like I said, if you have 100 or more articles in one category, it is **impossible** to find anything.

**Comment by squidfunk** [comment]
Okay, so maybe I have misunderstood you then. Language barrier. Would pagination on a category index page help? Category and archive index pages are currently not paginated, because I haven't had the case with 100 articles per category.

**Comment by ZhiyuanChen** [comment]
> Okay, so maybe I have misunderstood you then. Language barrier. Would pagination on a category index page help? Category and archive index pages are currently not paginated, because I haven't had the case with 100 articles per category.


Sorry I didn't make it clear
I have been using MkDocs for my personal blogs for years... and there are a lot of articles now (maybe not 100, but enough to make it difficult to find a specific article

Prior to blog plugin, I used to have nested sections for articles and it had worked great.
I just really like the idea to list all all articles in index page, and really like to have my name/avatar display next to my article.

I think pagination is not very necessary.
Most of the time, the users will not scroll at all if they didn't find anything interesting at first glance. 
So I'd really like something like the behavior without `navigation.section` in normal pages.

It could be even better if we have similar navigation inside each article. Left side underneath the author info or put `toc` on the left and put navigation on the right.

**Comment by squidfunk** [comment]
Thanks for the additional details, but I'm afraid I'm still not getting what you're aiming for. Could you maybe build a mockup, e.g. by editing some screenshots or explaining it differently? Is it nested categories? This was already asked for (can't find the issue right now, GitHub's issue search is not great).

**Comment by ZhiyuanChen** [comment]
1. Category page

Below figure demonstrate a category page

![image](https://user-images.githubusercontent.com/28757366/206903794-6b813b36-d675-4459-9bb4-ee912551cf6c.png)

Currently, the left side (denoted in blue) lists all categories and the right side (denoted in green) is empty. 

I think it would be nice to have something like a table of content on the right side that lists the name of articles in this category, something like this:

![image](https://user-images.githubusercontent.com/28757366/206904077-bd1834c3-6f1d-4009-bdc1-3cd322e9d35f.png)

**Comment by ZhiyuanChen** [comment]
2. Article Page

Below figure demonstrate an article page

![image](https://user-images.githubusercontent.com/28757366/206904198-21c3fcb0-7a84-41ea-8a8d-c7b7e49f942b.png)

Currently, the left side under author info (denoted in blue) is empty and the right side (denoted in green) is table of content.

I really hope there would be a "table of articles" in the left, or put table of content in the left and use the right spaces for "table of articles". So it should look like 

<img width="1162" alt="image" src="https://user-images.githubusercontent.com/28757366/206904796-ffe3be3f-ce7d-4824-ac65-bd0b4ee3fdd5.png">


or 

<img width="1178" alt="image" src="https://user-images.githubusercontent.com/28757366/206904537-9b6c2947-dec4-4057-97ac-424c7c1e68a2.png">

(personally I prefer the latter one as its more consistent with the category page and usually "table of articles" should be longer than table of contents

**Comment by ZhiyuanChen** [comment]
3. `post_url_format`

Currently `post_url_format` only supports `slug`, `file`, and `date`. It would be better if we can have the option to keep the original path or have `category` be part of the url.

**Comment by squidfunk** [comment]
Thanks for the input. What if multiple categories are present? Should only the first be part of the URL?

**Comment by squidfunk** [comment]
Phew, honestly, I'm not sure this is something many users want. So I can understand the use case with listing the articles in the table of contents on the index page, and maybe categories in URLs, but it seems that this may be something where different users have different needs, and I'm not sure we can generalize it enough so that it caters to all of them.

I'll investigate it the next time I work on the blog plugin.

**Comment by ZhiyuanChen** [comment]
> Phew, honestly, I'm not sure this is something many users want. So I can understand the use case with listing the articles in the table of contents on the index page, and maybe categories in URLs, but it seems that this may be something where different users have different needs, and I'm not sure we can generalize it enough so that it caters to all of them.
> 
> 
> 
> I'll investigate it the next time I work on the blog plugin.

Thank you so much, btw, is it possible to also support comment for blog posts?

**Comment by squidfunk** [comment]
Support for categories in post URLs was added in [`67ad8ba45`](https://github.com/squidfunk/mkdocs-material-insiders/commit/67ad8ba456a3caf608fb8de5a1d962bc510fc084). If [`post_url_format`](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#+blog.post_url_format) contains a `{categories}` placeholder, a posts' categories are slugified and added to the URL. By default, only the first category is added. If multiple categories are present, and `post_url_max_categories` is set to a number `> 1` (the default), they are joined with `/`.

__Example__:

In `mkdocs.yml`, add:

``` yaml
plugins:
  - blog:
      post_url_format: "{categories}/{slug}"
      post_url_max_categories: 2
```

If a post contains two categories `foo` and `bar`, the URL will be `foo/bar/post-title`.

**Comment by squidfunk** [comment]
Support for table of contents on category pages was added in [`1585c5e91`](https://github.com/squidfunk/mkdocs-material-insiders/commit/1585c5e91eba3cdbecbdd3e08cfb67047ab2e659). Enable it in `mkdocs.yml` with:

``` yaml
plugins:
  - blog:
      categories_toc: true
```

It looks like this:

![screenshot-localhost-8000-mkdocs-material-blog-category-search-1670880494920](https://user-images.githubusercontent.com/932156/207158352-34e0e16f-bbe2-488c-be52-c980d1204434.png)

**Comment by squidfunk** [comment]
With those changes, I'd consider this as completed. The prev/next pages on the category pages link to the previous and next category page, i.e. the previous or next page in the navigation tree. This behavior is not configurable, since category index pages are just single pages.

**Comment by squidfunk** [comment]
Released as part of [Insiders 4.27.0](https://squidfunk.github.io/mkdocs-material/reference/#built-in-typeset-plugin)!
